Adding new TRR for pass-the-ticket attack (#5)
tired-labs/techniques: reports/trr0000/win/README.md

## Metadata

| Key          | Value                                      |
|--------------|--------------------------------------------|
| ID           | TRR0000                                    |
| External IDs | [T1550.003]                                |
| Tactics      | Defense Evasion, Lateral Movement          |
| Platforms    | Windows                                    |
| Contributors | Andrew VanVleet                            |

### Scope Statement

There are many ways for an attacker to acquire Kerberos tickets. While this TRR
will discuss the most common methods used in a pass-the-ticket attack, it is not
the author's intent to make an exhaustive review of all the means by which an
adversary could acquire a valid Kerberos ticket. This TRR will primarily focus
on how an attacker can inject a previously-acquired ticket into their Kerberos
ticket cache, enabling them to assume the identity of the ticket owner.

## Technique Overview

Kerberos is a common authentication protocol supported natively by all modern
operating systems. It is also used extensively for authentication and
authorization in Active Directory domains. As such, it plays a role in probably

> every modern enterprise network. An attacker who can steal valid Kerberos

Reviewer: ```suggestion
authorization in Active Directory domains. As such, it plays a role in many
modern enterprise networks. An attacker who can steal valid Kerberos
```

You know i'm a negative nelly when it comes to words that introduce doubt such as "probably". Suggesting a different way to phrase this

tickets from other users can use this procedure to inject those tickets into
their own current logon session and thereby assume the identity of the ticket's
owner.

The Windows operating system has a built-in security package that implements the
Kerberos protocol (found in the library `System32\kerberos.dll`). This security
package supports the ability to add and retrieve tickets from a local cache as a
fundamental part of its Kerberos implementation. Attackers can manipulate this
functionality to acquire and insert tickets belonging to other users.

## Technical Background

The key concepts required to understand this attack are Kerberos tickets and the
Windows implementation of the protocol.

### Kerberos tickets

Kerberos is a network authentication protocol designed to provide strong
authentication for client/server applications by using symmetric key

> cryptography to sign 'tickets' that a client can use to prove their identity and

Reviewer 2: Is there a standard or precedent for single vs double quotes? That's probably too obsessive-compulsive, but it's what my brain immediately thought. Feel free to ignore this.

provide authorization to access a network resource.

Kerberos relies on a Key Distribution Center (KDC), comprised of 3 elements.
While these can be distributed over multiple servers, most modern
implementations have them on the same server. These elements are:

- An authentication server (AS) that performs the initial authentication.
- A ticket-granting server (TGS) that issues tickets to clients for specific
  services hosted on APs.
- A Kerberos database that stores the password hash and identity of all verified
users. 

The Kerberos protocol defines a series of messages exchanged between the

> client, the KDC and AP. After a client has authenticated, they receive a Ticket

Reviewer 3: ```suggestion
client, the KDC, and AP. After a client has authenticated, they receive a Ticket
```

I am #TeamOxfordComma

can then provide the service ticket to the AP to gain access to the service.

Kerberos is stateless; it relies on cryptographically signed tickets for
authentication and authorization. This requires that the cryptographic keys are
known only by the key's owner and the KDC. There are a few different keys used
for these cryptographic signatures:

- TGTs are signed with the password hash of the `KRBTGT` account, which is a
special service account that represents the KDC in the directory.
- Service tickets are signed with the hash of the service account associated
with a given service (a CIFS share, an FTP server, etc). In the directory,
services are represented by a service principal name (SPN) connected to the
associated service account. The password of the service account should be known
only by the KDC and the service.
- Clients prove their identity by signing a timestamp with their own password
hash, which should be known only by the KDC and themselves.

A pass-the-ticket attack abuses the statelessness of the Kerberos protocol. The
possessor of a valid ticket is implicitly accepted as a valid owner of that

> identity, because the protocol uses cryptography (specifically, session keys and

Reviewer 3: ```suggestion
identity because the protocol uses cryptography (specifically, session keys and
```

password hashes that should be known only to the owner) to ensure that tickets
are issued only to their legitimate owners. A pass-the-ticket attack abuses the
mechanisms for storing and retrieving tickets after they have been issued to the
verified owner.

### Windows Local Security Authority (LSA)

The Local Security Authority (LSA) is a protected Windows subsystem that manages
critical security functions like user authentication (validating credentials,
issuing access tokens, etc), enforcement of security policies (like password
complexity requirements and account lockouts), and securely storing secrets
(password hashes, Kerberos tickets, etc). The LSA Subsystem Service (LSASS) is
the process that implements the LSA's functionality; this runs as the trusted
`SYSTEM` account in a process named `lsass.exe.` The LSASS process is a rich
target for attackers, so it has many built-in protections like [Credential
Guard] and [Protected Processes].

> 

Reviewer 4: Potentially expand on how Credential guard attempts to protect lsass. Completely understand if this is out of scope, but it does change how lsass.exe presents to the end user.  Something like this blurb Found here: https://learn.microsoft.com/en-us/windows/security/identity-protection/credential-guard/: "Credential Guard uses [Virtualization-based security (VBS)](https://learn.microsoft.com/en-us/windows-hardware/design/device-experiences/oem-vbs) to isolate secrets so that only privileged system software can access them. Unauthorized access to these secrets can lead to credential theft attacks like pass the hash and pass the ticket.

When enabled, Credential Guard provides the following benefits:

    Hardware security: NTLM, Kerberos, and Credential Manager take advantage of platform security features, including Secure Boot and virtualization, to protect credentials
    Virtualization-based security: NTLM, Kerberos derived credentials, and other secrets run in a protected environment that is isolated from the running operating system
    Protection against advanced persistent threats: when credentials are protected using VBS, the credential theft attack techniques and tools used in many targeted attacks are blocked. Malware running in the operating system with administrative privileges can't extract secrets that are protected by VBS"

Reviewer 3: To add to that, `lsass.exe` is mentioned above, but if Cred Guard is enabled, the creds are actually in lsaiso.exe (although there is still an lsass.exe that I believe is just a "proxy"). I don't know if that clarification is worth including, but just wanted to mention it.

Reviewer: I think that strikes the perfect balance. Well done!

provides a common interface regardless of the SSP used.) However, it is possible
to directly invoke Kerberos SSP/AP functionality using the
[LsaCallAuthenticationPackage] API. Here is the function definition:

```code
NTSTATUS LsaCallAuthenticationPackage(
  [in]  HANDLE    LsaHandle,
  [in]  ULONG     AuthenticationPackage,
  [in]  PVOID     ProtocolSubmitBuffer,
  [in]  ULONG     SubmitBufferLength,
  [out] PVOID     *ProtocolReturnBuffer,
  [out] PULONG    ReturnBufferLength,
  [out] PNTSTATUS ProtocolStatus
);
```

The `ProtocolSubmitBuffer` parameter is used to submit a protocol-specific
message to the security package hosted in LSASS. The Kerberos SSP/AP supports
[just short of 40 messages]. The ones relevant to this attack technique are:

- `KerbQueryTicketCacheMessage` - queries the local ticket cache
- `KerbRetrieveEncodedTicketMessage` - retrieves a ticket from the local
  cache
- `KerbSubmitTicketMessage` - submits a ticket to be included in the local cache

> [!NOTE]
> This list of messages is not necessarily inclusive. It's probably possible to
> accomplish the same task with some of the other supported messages, like

> > `KerbQueryTicketCacheExMessage` or `KerbAddExtraCredentialsMessage`.

Reviewer 5: Can I decline on this suggestion?  Seems like a wordier way to say the same thing.  Sometimes you can't avoid statements of likelihood. ;)

|-----------------|--------------------------------|-------------------|
| TRR0000.WIN.A   | Submit ticket to local cache | Defense Evasion, Lateral Movement |

### Procedure A: Submit ticket to local cache

Once a ticket has been stolen, it can be injected into any logon session using
the Kerberos SSP/AP `KerbSubmitTicketMessage` message. This submits the ticket
to be included in the current logon session's ticket cache, where it will be
used whenever access is requested. One important detail to note here: it is
pointless for an attacker to inject a ticket for a user into a logon session for
that same user, because it's a simple matter to request a new valid ticket for
the logged on user. **Thus, maliciously injected tickets will always have a
mismatch between the client the ticket was issued for and the logon session user
account.**

#### Detection Data Model

![DDM - Procedure Name 1](ddms/ddm_trr0000_win_a.png)

#### A Note on Ticket Acquisition

As noted in the scope statement, there are many different ways an attacker could
potentially acquire valid Kerberos tickets that could be used in a
pass-the-ticket attack. The DDM includes the ones most commonly used with this

> procedure, like those used by the popular tools [Rubeus] and [Mimikatz]. The DDM

Reviewer: ```suggestion
pass-the-ticket attack. For reference, the DDM includes the ones most commonly seen by the author, primarily used by the popular tools [Rubeus] and [Mimikatz]. The DDM
```

also acknowledges that there are other ways, but this TRR will not attempt an

> exhaustive exploration of them. All methods for acquiring tickets are in gray on

Reviewer 2: ```suggestion
also acknowledges that there are other ways, but this TRR will not attempt an
exhaustive exploration of them. All methods in gray are for acquiring tickets on
```
"not exhaustive... All methods...". I knew what you meant, but it was a little unclear grammatically.

the DDM, in addition to the operation for the original valid request for the
ticket, to note that these actions are precursors that must occur before the
ticket can be imported into a new logon session.

> [!NOTE]
>
> The process of reading (or dumping) LSASS process memory to extract cached
> Kerberos tickets is functionally equivalent to attack technique [T1003.001] OS
> Credential Dumping: LSASS Memory. Controls to detect or prevent T1003.001 will
> also detect or prevent an attacker from acquiring tickets from LSASS.

## Available Emulation Tests

| ID            | Link             |
|---------------|------------------|

> | TRR0000.WIN.A | [Atomic Tests 1-2]          |

Reviewer 5: It's not filler junk, those are the links to the Atomic Red Team tests that can be used to emulate a PTT attack.
